hibernate-core/src/main/java/org/hibernate/query/sqm/internal/TypecheckUtil.java

```diff
@@ -392,6 +395,23 @@ private static Class<?> canonicalize(Class<?> lhs) {
 		};
 	}
 
+	private static boolean isMappedSuperclassTypeAssignable(
+			MappedSuperclassDomainType<?> lhsType,
+			EntityType<?> rhsType,
+			BindingContext bindingContext) {
+
+		for ( ManagedDomainType<?> candidate : lhsType.getSubTypes() ) {
+			if ( candidate instanceof MappedSuperclassDomainType<?> candidateMappedSuperclass  &&
+					isMappedSuperclassTypeAssignable( candidateMappedSuperclass, rhsType, bindingContext )) {
+				return true;
+			}
+			if ( candidate instanceof EntityType<?> candidateEntityType &&
+					isEntityTypeAssignable(candidateEntityType, rhsType, bindingContext) ) {
+				return true;
+			}
+		}
+		return false;
+	}
```

Comment:
This is much better, just a couple of stylistic changes (I would intuitively put the entity-type check first, not that it makes a big difference).

```suggestion
		for ( ManagedDomainType<?> candidate : lhsType.getSubTypes() ) {
			if ( candidate instanceof EntityType<?> candidateEntityType &&
					isEntityTypeAssignable( candidateEntityType, rhsType, bindingContext ) ) {
				return true;
			}
			else if ( candidate instanceof MappedSuperclassDomainType<?> candidateMappedSuperclass &&
					isMappedSuperclassTypeAssignable( candidateMappedSuperclass, rhsType, bindingContext ) ) {
				return true;
			}
		
		}
		return false;
	}

```